modules/indexing/src/main/java/org/apache/ignite/internal/processors/query/stat/IgniteStatisticsManagerImpl.java

```diff
@@ -0,0 +1,232 @@
+        statsRepos.clearGlobalStatistics(key, colNames);
+    }
+
+    /**
+     * Filter columns by specified names.
+     *
+     * @param columns columns to filter.
+     * @param colNames names.
+     * @return column with specified names.
+     */
+    private Column[] filterColumns(Column[] columns, String... colNames) {
+        if (colNames == null || colNames.length == 0) {
+            return columns;
+        }
+        List<Column> resultList = new ArrayList<>(colNames.length);
+
+        for (String colName : colNames)
+            for (Column col : columns)
+
+                if (colName.equals(col.getName())) {
+                    resultList.add(col);
+                    break;
+                }
+
+        return resultList.toArray(new Column[resultList.size()]);
+    }
+
+    @Override public void collectObjectStatistics(String schemaName, String objName, String... colNames)
+            throws IgniteCheckedException {
+        GridH2Table tbl = schemaMgr.dataTable(schemaName, objName);
+        if (tbl == null)
+            throw new IgniteAuthenticationException(String.format("Can't find table %s.%s", schemaName, objName));
+
+        if (log.isDebugEnabled())
+            log.debug(String.format("Starting statistics collection by %s.%s object", schemaName, objName));
+
+        Column[] selectedColumns;
+        boolean fullStat;
+        if (colNames == null || colNames.length == 0) {
+            fullStat = true;
+            selectedColumns = tbl.getColumns();
+        } else {
+            fullStat = false;
+            selectedColumns = filterColumns(tbl.getColumns(), colNames);
+        }
+
+        Collection<ObjectPartitionStatisticsImpl> partsStats = collectPartitionStatistics(tbl, selectedColumns);
+        StatsKey key = new StatsKey(tbl.identifier().schema(), tbl.identifier().table());
+        statsRepos.saveLocalPartitionsStatistics(key, partsStats, fullStat);
+
+        ObjectStatisticsImpl tblStats = aggregateLocalStatistics(tbl, selectedColumns, partsStats);
+        statsRepos.saveLocalStatistics(key, tblStats, fullStat);
+        if (log.isDebugEnabled())
+            log.debug(String.format("Statistics collection by %s.%s object is finished.", schemaName, objName));
+    }
+
+    private Collection<ObjectPartitionStatisticsImpl> collectPartitionStatistics(GridH2Table tbl, Column[] selectedColumns)
+            throws IgniteCheckedException {
+        List<ObjectPartitionStatisticsImpl> tblPartStats = new ArrayList<>();
+        GridH2RowDescriptor desc = tbl.rowDescriptor();
+        String tblName = tbl.getName();
+
+        for (GridDhtLocalPartition locPart : tbl.cacheContext().topology().localPartitions()) {
+            final boolean reserved = locPart.reserve();
+
+            try {
+                if (!reserved || (locPart.state() != OWNING && locPart.state() != MOVING)
+                        || !locPart.primary(ctx.discovery().topologyVersionEx()))
```

Comment:
Do we really want to gather statictics on partially moved partition?
Actually, primary can't be in moving state.
Discovery component gives you latest observed topology in grid, but assignment can be NOT ready yet.
I'd think you can get lastTopologyReady version from exchange(), not discovery().